### PR TITLE
Adds Json Parsing to nested object during update Query step in ML Inference Request processor

### LINKED ISSUE #3767: [Enhancement] Support nested object substitution in query_template within ML inference Request Processor
**Is your feature request related to a problem?**
A clear and concise description of what the problem is, e.g. _I'm always frustrated when [...]_

**What solution would you like?**

**What alternatives have you considered?**

**Do you have any additional context?**

### PR DESCRIPTION
**Pull request by brianf-aws**
### Description
Ensures that Sparse Vectors can be parsed by ML Inference Request processor during rewrite.

Next steps can be to write a tutorial showing how we can run neural sparse search with ML Inference request processor.

### Related Issues
Resolves #3767

### Testing
spotlessApply && `./gradlew test`

### Check List
- [X] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [X] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java [hunk 1]
```diff
@@ -46,6 +46,7 @@
 import org.opensearch.search.pipeline.SearchRequestProcessor;
 import org.opensearch.transport.client.Client;
 
+import com.google.gson.Gson;
```
Comment (mingshl) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:49]: can you use the toJson from StringUtils here. https://github.com/opensearch-project/ml-commons/blob/ade09e65639aaa83e68fdc2db00f04c35c0a5e19/common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java#L180 ? trying to organize the import.

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java [hunk 2]
```diff
@@ -360,6 +361,9 @@ private String updateQueryTemplate(String queryTemplate, Map<String, String> out
                     String newQueryField = outputMapEntry.getKey();
                     String modelOutputFieldName = outputMapEntry.getValue();
                     Object modelOutputValue = getModelOutputValue(mlOutput, modelOutputFieldName, ignoreMissing, fullResponsePath);
+                    if (modelOutputValue instanceof Map) {
+                        modelOutputValue = new Gson().toJson(modelOutputValue);
+                    }
```
Comment (mingshl) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:366]: this is the right place to fix the issue, the string convert to json string should happen before string substitution. So the next question is, should this toJson conversion only added when it's a map?

**Comment by mingshl** [comment]
Thanks @brianf-aws for taking up the fix!!! overall it's a pretty neat fix, leave some comments.

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java (hunk 1 above)
Comment (brianf-aws) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:49]: Good callout will add this in my next commit

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java (hunk 2 above)
Comment (brianf-aws) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:366]: Map should be a catch all for many scenarios that I can think of at the top of my head. 

I may need to test if this solution doesn't break the query from remote llm. I wonder if a list should also be supported as well for example a list of maps. then this instanceof check wouldnt work in that scenario.

Comment (brianf-aws) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:365]: I made a UT where a llm returns a query and a query rewrite occurs. This happened to succeed with the code change e.g.
```
        /**
         * {
         *   "inference_results" : [ {
         *     "output" : [ {
         *       "name" : "response",
         *       "dataAsMap" : {
         * 		  "content": [
         * 			"text": "{\"query\": \"match_all\" : {}}"
         *       }
         *     } ]
         *   } ]
         * }
         */
```

Off the top of my head I cant think of another scenario where this parsing issue occurs since I see that your UTs cover a lot of edge cases such as a list.

File: plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessorTests.java [hunk 3]
```diff
@@ -1499,6 +1504,216 @@ public void onFailure(Exception e) {
...
+                TEST_XCONTENT_REGISTRY_FOR_QUERY
+        );
+
+        /**
+         * {
+         *   "inference_results" : [ {
+         *     "output" : [ {
+         *       "name" : "response",
+         *       "dataAsMap" : {
+         * 		  "content": [
```
Comment (brianf-aws) [inline plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessorTests.java:1676]: Not sure why its displayed out of place on github but looks normal in intelij

**Comment by brianf-aws** [comment]
> Spotless failed, let's apply spotless.

Ah! okay have fixed can you take a look @jngz-es ?

**Review by Zhangxunmt — APPROVED** [review]

**Comment by jngz-es** [comment]
Another error.
```
* What went wrong:
Execution failed for task ':opensearch-ml-plugin:forbiddenPatterns'.
> Found invalid patterns:
  - tab on line 1671 of plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessorTests.java
  - tab on line 1672 of plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessorTests.java
 ```

**Comment by brianf-aws** [comment]
Hey @jngz-es I fixed the issue

you can see locally the test passing
```
./gradlew opensearch-ml-plugin:forbiddenPattern                 
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 8.11.1
  OS Info               : Mac OS X 15.4.1 (aarch64)
  JDK Version           : 21 (Amazon Corretto JDK)
  JAVA_HOME             : /Users/<>/.sdkman/candidates/java/21.0.5-amzn
  Random Testing Seed   : 9B2C658F659F0CDD
  In FIPS 140 mode      : false
=======================================
[Incubating] Problems report is available at: file:///Users/<>/IdeaProjects/ml-commons/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.11.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 556ms

```

I'm guessing tabs are prohibited when writing javadocs

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java (hunk 2 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:365]: we should cover more use cases in testing, because the model output can be various format, 

I have two cases in mind that you can try add and verify in the UT 
1. list of map, 
```
{
  "query": {
    "bool": {
      "must": [
        {
          "terms": {
            "categories": [
              {"name": "electronics", "id": 1},
              {"name": "computers", "id": 2},
              {"name": "phones", "id": 3}
            ]
          }
        }
      ]
    }
  }
}
```
2. list of nested map (map of map), 
```
{
  "query": {
    "bool": {
      "must": [
        {
          "nested": {
            "path": "products",
            "query": {
              "bool": {
                "must": [
                  {
                    "terms": {
                      "products.variants": [
                        {
                          "color": {
                            "primary": {"hex": "#FF0000", "name": "red"},
                            "secondary": {"hex": "#000000", "name": "black"}
                          },
                          "size": {
                            "dimensions": {"width": 10, "height": 20},
                            "label": {"us": "M", "eu": "38"}
                          }
                        },
                        {
                          "color": {
                            "primary": {"hex": "#0000FF", "name": "blue"},
                            "secondary": {"hex": "#FFFFFF", "name": "white"}
                          },
                          "size": {
                            "dimensions": {"width": 12, "height": 22},
                            "label": {"us": "L", "eu": "40"}
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            }
          }
        }
      ]
    }
  }
}
```

Comment (brianf-aws) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:365]: For Scenario one list of Map I dont think this is a valid OpenSearch query consequently If it runs through the processor it fails the query rewrite

However I did try this, taken from [docs](https://arc.net/l/quote/clprrqsi) :

```
query_template=
{
  "query": {
    "bool": {
      "must": [
       ${llm_response}
      ]
    }
  }
}


llm_response = 
 {
          "bool": {
            "should": [
              {
                "match": {
                  "text_entry": "love"
                }
              },
              {
                "match": {
                  "text": "hate"
                }
              }
            ]
          }
        },
        {
          "bool": {
            "should": [
              {
                "match": {
                  "text_entry": "life"
                }
              },
              {
                "match": {
                  "text": "grace"
                }
              }
            ]
          }
        }
```

**Comment by mingshl** [comment]
overall it;s looking good, would you increase the codecov?

**Comment by brianf-aws** [comment]
> overall it;s looking good, would you increase the codecov?

I only edited one file and introduced one branch. But I did rebase recently so I assume the code coverage extension is judging me based on that too.

**Comment by dhrubo-os** [comment]
> > overall it;s looking good, would you increase the codecov?
> 
> I only edited one file and introduced one branch. But I did rebase recently so I assume the code coverage extension is judging me based on that too.

I think we should be looking at this: https://github.com/opensearch-project/ml-commons/pull/3856/checks?check_run_id=42675669550

We should have a delta coverage report like [this](https://github.com/opensearch-project/flow-framework/blob/ef4d2443ad644e520a2a18d4cf13fbd266a193fc/build.gradle#L752-L770)


@brianf-aws could you please create an issue on this? Thanks

**Review by mingshl — APPROVED** [review]
the code coverage for this change is 100%, approved

**Comment by brianf-aws** [comment]
> I think we should be looking at this: https://github.com/opensearch-project/ml-commons/pull/3856/checks?check_run_id=42675669550
> 
> We should have a delta coverage report like [this](https://github.com/opensearch-project/flow-framework/blob/ef4d2443ad644e520a2a18d4cf13fbd266a193fc/build.gradle#L752-L770)
> 
> @brianf-aws could you please create an issue on this? Thanks

Hey @dhrubo-os  I created a issue to track : https://github.com/opensearch-project/ml-commons/issues/3866

**Review by dhrubo-os — APPROVED** [review]
